Allow people to add routing labels with commands

grodrigues3 opened the pull request:
I'm open to changing the command name but does the functionality and design sound okay.

As currently implemented you could type

/label sig-testing sig-network area/kubelet 

And all of those labels would be added.

Fixes #1640

spxtr commented:
As implemented, anyone in the org can type `/label lgtm approved`, which is not something we should allow.

apelisse commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+
+var (
+	COMMAND_NAME    = "label"
+	labelRegex      = regexp.MustCompile(`(?m)^/(` + COMMAND_NAME + `)[\t ]*([^\n\r]*)`)
```
I can't believe every plugin has to do that.

spxtr replied:
We can change to a better way if we want. It's not too terrible the way it is, but it's certainly possible to introduce slight differences that might confuse people.

apelisse commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+
+func handle(gc githubClient, log *logrus.Entry, ic github.IssueCommentEvent) error {
+	// Only consider open PRs.
+	if !ic.Issue.IsPullRequest() || ic.Issue.State != "open" || ic.Action != "created" {
```
why only pull-requests?

spxtr replied:
Why only open issues? I might want to label a closed PR :)

spxtr commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+limitations under the License.
+*/
+
+package label
```
Name the file `label.go`.

spxtr commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+const pluginName = "add_label"
+
+var (
+	COMMAND_NAME    = "label"
```
It's unusual to see `CAPITALIZED_CONSTANTS` in go. Since this is only used once, I would just remove it and make the next line easier to read.

spxtr commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+	repo := ic.Repo.Name
+	number := ic.Issue.Number
+
+	val, err := gc.IsMember(repo, login)
```
```go
member, err := gc.IsMember(repo, login)
if err != nil {
	return fmt.Errorf("IsMember failed: %v", err)
}
if !member {
	log.Info("Commenting that only members can add labels.")
	return gc.CreateComment(...)
}
```

spxtr commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+var (
+	COMMAND_NAME    = "label"
+	labelRegex      = regexp.MustCompile(`(?m)^/(` + COMMAND_NAME + `)[\t ]*([^\n\r]*)`)
+	AllowedPrefixes = sets.NewString("sig", "area")
```
Go has sets of strings already. `var allowedPrefixes map[string]bool`.

spxtr replied:
Since you're just going to iterate over it anyway, probably just `[]string` is fine.

grodrigues3 replied:
I want it to be easy to identify the actual command name.  If the regex looks like

``(?m)^/(label)[\t ]*([^\n\r]*)``

It might be harder to identify that the command name is `label`

spxtr commented on prow/plugins/label/add_label.go:
```diff
@@ -0,0 +1,87 @@
+	for _, match := range matches {
+		newLabel := strings.TrimSpace(match[2])
+		if isAllowed(newLabel) {
+			log.Infof("Error adding label %v to issue %v: %v", newLabel, number, err)
```
Add the label :)

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,87 @@
+const pluginName = "add_label"
+
+var (
+	COMMAND_NAME    = "label"
```
`const commandName = "label"`

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,87 @@
+	for _, match := range matches {
+		newLabel := strings.TrimSpace(match[2])
+		if isAllowed(newLabel) {
+			log.Infof("Error adding label %v to issue %v: %v", newLabel, number, err)
```
I think the logic here is incorrect.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,87 @@
+	repo := ic.Repo.Name
+	number := ic.Issue.Number
+
+	member, err := gc.IsMember(repo, login)
```
I think `ic.Repo.Owner.Login` is the org name, such as `kubernetes`. `ic.repo.Name` would be something like `test-infra`. You want to check that the comment author is in the org.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,87 @@
+	}
+	if !member {
+		log.Info("Commenting that only members can add labels.")
+		return gc.CreateComment(login, repo, number, "Only org members can add labels")
```
Use `plugins.FormatResponse`: 
```
return gc.CreateComment(org, repo, number, plugins.FormatResponse(ic.Comment, "Only kubernetes org members can add labels."))
```

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,87 @@
+
+var (
+	COMMAND_NAME    = "label"
+	labelRegex      = regexp.MustCompile(`(?m)^/(` + COMMAND_NAME + `)[\t ]*([^\n\r]*)`)
```
Why not `(?m)^/label ([\w/-]*)\r?$`?

spxtr commented:
Two broader questions:

1. Should `/label area/foo sig-bla` work? I think it should.
1. If it already has `area/foo`, should `/label area/foo` remove the label? I tend to think so, but I'm not sure.

grodrigues3 commented:
1. Totally agree.  Slashes and dashes most definitely should be included.

1. I think two separate commands might be better.  /label and /unlabel or something like that?  I also thought i'd start testing the adding functionality first and collect feedback.  

@apelisse WDYT?

apelisse commented:
Oh I just realized I didn't send my message ...

Yes, I think 1. is about having multiple labels on the same command, and you're working on this. If it's actually about dash and slash, sure, should work too ...
2. I would wait to see need, and I would make it into two commands.

grodrigues3 commented:
@spxtr I think I addressed your comments.  There are tests that cover valid labels and invalid labels (e.g. lgtm).  Multiple labels in one argument and the command used multiple times on different lines.  Build files are updated.  Let me know if there's anything else.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,91 @@
+		return fmt.Errorf("IsMember failed: %v", err)
+	}
+	if !member {
+		return gc.CreateComment(owner, repo, number, plugins.FormatResponse(ic.Comment, "Only kubernetes org members can add labels."))
```
Add a `log.Infof` saying that you're commenting.

grodrigues3 replied:
gc.CreateComment logs before creating the log I think

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,91 @@
+	for _, newLabel := range strings.Split(match[0], " ")[1:] {
+		newLabel = strings.TrimSpace(newLabel)
+		if isAllowed(newLabel) {
+			gc.AddLabel(owner, repo, number, newLabel)
```
Add a `log.Infof` saying that you're adding the label.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,91 @@
+		if isAllowed(newLabel) {
+			gc.AddLabel(owner, repo, number, newLabel)
+		} else {
+			gc.CreateComment(owner, repo, number, fmt.Sprintf(LabelNotAllowedMsg, newLabel, number))
```
Add a `log.Infof` saying that you're commenting.

spxtr replied:
Maybe accumulate the list of labels that are disallowed and make one comment at the end? Right now if I do `/label approve lgtm somethingelse` I'll get 3 comments, when really it can be one.

grodrigues3 replied:
The first line in CreateComment is to log.  I didn't want to log 2x

grodrigues3 replied:
Fair point on accumulated.  I'll update

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,91 @@
+		return nil
+	}
+	for _, newLabel := range strings.Split(match[0], " ")[1:] {
+		newLabel = strings.TrimSpace(newLabel)
```
I think we want to ignore the command if the label is already on the issue.
```
if ic.Issue.HasLabel(newLabel) {
	continue
}
```

grodrigues3 commented:
Excellent point on the naming confusion.  Let me think about it a bit more before getting it merged (i'll also update commands.md at that time).

grodrigues3 commented:
@spxtr @ixdy pretty ambiguous build failed log.  any advice?

spxtr commented:
@k8s-bot bazel test this

spxtr commented:
Please fix the merge conflict. We should still be getting results, though. I'm seeing `W0301 17:33:28.154] ResumableUploadAbortException: 403 Insufficient Permission`, which is worrying.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,96 @@
+const pluginName = "label"
+
+var (
+	AllowedPrefixes    = []string{"sig", "area", "kind"}
```
nit: `allowedPrefixes`

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,96 @@
+var (
+	AllowedPrefixes    = []string{"sig", "area", "kind"}
+	labelRegex         = regexp.MustCompile(`(?m)^/label[\t\s]+(?:\S*(?:\s+|$))*`)
+	LabelNotAllowedMsg = "The following labels: %v are not allowed to be added with the label command"
```
nit: `labelNotAllowedMsg`

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func handle(gc githubClient, log *logrus.Entry, ic github.IssueCommentEvent) error {
+
```
nit: remove newline

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,96 @@
+	if match == nil || len(match) == 0 {
+		return nil
+	}
+	failures := []string{}
```
nit: `var failures []string`

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,96 @@
+	failures := []string{}
+	for _, newLabel := range strings.Split(match[0], " ")[1:] {
+		newLabel = strings.TrimSpace(newLabel)
+		if isAllowed(newLabel) {
```
I think we want to ignore the command if the label is already on the issue.
```
if ic.Issue.HasLabel(newLabel) {
	continue
}
```

spxtr commented on prow/plugins/label/label_test.go:
```diff
@@ -0,0 +1,142 @@
+)
+
+const (
+	FakeRepoOrg  = "fakeOrg"
```
nit: `fakeRepoOrg` and `fakeRepoName`.

grodrigues3 commented:
@k8s-bot bazel test this

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,98 @@
+var (
+	allowedPrefixes    = []string{"sig", "area", "kind"}
+	labelRegex         = regexp.MustCompile(`(?m)^/label[\t\s]+(?:\S*(?:\s+|$))*`)
+	labelNotAllowedMsg = "The following labels: %v are not allowed to be added with the label command"
```
As discussed, this comment should say which labels are allowed.

grodrigues3 replied:
yup, i'll look at this tomorrow

apelisse commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+
+var (
+	allowedPrefixes = []string{"area", "priority"}
+	regexBuilder    = `(?m)^/%s[\t\s]+(?:\S*(?:[^\n]+))*`
```
Interestingly, you don't have any capture group. Don't you want to capture the argument list?

`(?:[^\n]+)` doesn't have to be a group at all. Consider replacing with `[^\n]+`

The `%s` is kind of gross, can you replace by the actual command?

I know that you have multiple command, and I see you're having a for loop over the list of prefixes, maybe you should replace directly in the regex?

`/(area|priority) ...`.

apelisse replied:
If this regex does what you want, it's not very clear why.

`[\t\s]+`: you need at least a space after the command, what about `/area$` ?

Can you explain this part: `(?:\S*(?:[^\n]+))*`

0 or more non-space, I assume this is the label, it can be empty?
Followed by non-`\n` characters, and this must be non-empty? What is this supposed to match?

grodrigues3 replied:
I think I'd prefer to capture each label individual, but I don't think you can write a regex to match and capture a variable number of groups (variable capture groups doesn't work).  The `[^\n]` was to avoid the multiple line problem (`/priority priority/critical\n/area area/a1`) will otherwise result in a match `priority/critical\n/area`

grodrigues3 replied:
`(?m)^/(area|priority)[\t\s]+(?:\S+(?:[^\n]+))*` something like this doesn't match multiple lines (because of the [\n])

apelisse commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+				continue
+			}
+			if strings.HasPrefix(newLabel, pref) {
+				if err := gc.AddLabel(owner, repo, number, newLabel); err != nil {
```
Why do you add labels one by one?

grodrigues3 replied:
The github client in prow doesn't yet support adding multiple at the same time

grodrigues3 replied:
because there's no function to add multiple at the same time.  also, i'm not sure what github does if some labels are valid and others aren't.

apelisse commented on prow/plugins/label/label_test.go:
```diff
@@ -0,0 +1,176 @@
+var areaLabels = []string{"area/api", "area/infra"}
+var priorityLabels = []string{"priority/important", "priority/critical"}
+
+func formatLabels(labels sets.String) sets.String {
```
you could make this take a `labels ...string` parameter instead.

apelisse replied:
Plus I have doubts that @spxtr will be happy to see the sets imported ;-)

spxtr replied:
I'm okay with importing it in a test file. Go has string sets built-in, `map[string]bool`, and while I think it's better to use that I'm not going to be *that* picky.

spxtr replied:
Note that sets doesn't import the whole of k8s apimachinery.

apelisse replied:
Yeah, didn't mean to offend you :-)
I do believe it will be much easier to read with `...string` though.

spxtr replied:
Agreed

apelisse commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+		}
+	}
+
+	if len(failures) > 0 {
```
it might fail because of github, it might fail because the format is invalid, it's not very explicit for the user.

apelisse commented on prow/plugins/label/label_test.go:
```diff
@@ -0,0 +1,176 @@
+		{
+			name:              "Empty Area",
+			body:              "/area",
+			expectedNewLabels: formatLabels(sets.NewString()),
```
You're test doesn't really detect failures :-/ I wish we knew what happens here.

apelisse commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+			if ic.Issue.HasLabel(newLabel) {
+				continue
+			}
+			if strings.HasPrefix(newLabel, pref) {
```
if the label doesn't start with the prefix, can we just add it? like `/priority critical` adds label `priority/critical`?

grodrigues3 replied:
how much fuzzy matching should we allow then? if it's just without the prefix that's fine with me, but I don't want to try and guess what the commenter meant

apelisse replied:
I agree, I kind of hate "do-what-I-mean" type of logic. Is it here though? I have the feeling that the prefix is 100% matching the name of the command?

grodrigues3 replied:
yeah, if that's going to be the convention i'm fine with that too.  but it should be `/priority critical-urgent` (matching fully)

apelisse replied:
Yeah, thanks for being thorough. At least I was self-consistent. But completely agree: it should match the exact label name.

grodrigues3 commented on prow/plugins/label/label_test.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func TestLabel(t *testing.T) {
+	// "a" is the author, "a", "r1", and "r2" are reviewers.
```
@apelisse @spxtr I need help figuring out a new testing strategy.  Before I'd compare fc.AddedLabels to expectedLabels but since I'm using fc.AddedLabels as the list of all labels (see changes to fakeClient where I added GetLabels func) I can no longer make that comparison.  Can I modify the functionality of handle so that it returns the labels that are successfully added to the issue?

spxtr replied:
You don't need to use `fakegithub.FakeClient`, feel free to write your own for individual tests.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,93 @@
+		return gc.CreateComment(owner, repo, number, plugins.FormatResponse(ic.Comment, "Only kubernetes org members can add labels."))
+	}
+
+	labels, err := gc.GetLabels(owner, repo)
```
IssueCommentEvents should have labels included already.

grodrigues3 replied:
I want all the labels that exist in the repo.  That doesn't exist in IssueCommentEvents, right?

spxtr replied:
Oh, no you're right.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+	repo := ic.Repo.Name
+	number := ic.Issue.Number
+
+	member, err := gc.IsMember(repo, commenter)
```
Check that it matches the regex before doing any API calls. As is, if any non-org member makes ANY comment, it will comment back.

spxtr commented on prow/plugins/label/label_test.go:
```diff
@@ -0,0 +1,190 @@
+		body              string
+		commenter         string
+		expectedNewLabels []string
+		repoLabels        []string
```
I think every case uses `existingLabels`?

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+		for _, newLabel := range strings.Split(match[0], " ")[1:] {
+			newLabel = match[1] + "/" + strings.TrimSpace(newLabel)
+			if ic.Issue.HasLabel(newLabel) || !existingLabels.Has(newLabel) {
+				continue
```
If I say `/area nonexistant` then I want it to comment back saying `There's no area/nonexistant label.` or something, not ignore it silently.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+const pluginName = "label"
+
+var (
+	labelRegex  = regexp.MustCompile(`(?m)^/(area|priority)\s*(.*)$`)
```
also `kind`?

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+			if ic.Issue.HasLabel(newLabel) || !existingLabels.Has(newLabel) {
+				continue
+			}
+			if err := gc.AddLabel(owner, repo, number, newLabel); err != nil {
```
We can change the API call to `AddLabels` and have it take a `[]string` as input, which would probably simplify this.

apelisse replied:
Yeah, the problem is that github API is very poorly defined. What happens if it can't apply one label? Does it ... apply all others? apply some? apply none? It's also very unclear if/how it will tell you what/which failed.

spxtr replied:
Erick was playing with this for assignees recently in #2118, where he returns a special type of error if it fails to assign.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	if len(failures) > 0 {
+		gc.CreateComment(owner, repo, number, fmt.Sprintf(labelFailed, strings.Join(failures, ", "), number))
```
Don't ignore errors!

apelisse commented on prow/github/fakegithub/fakegithub.go:
```diff
@@ -103,9 +106,20 @@ func (f *FakeClient) CreateStatus(owner, repo, ref string, s github.Status) erro
+		return nil
+	}
+	for _, l := range f.ExistingLabels {
+		if label == l {
```
You forgot some code in this `if`

grodrigues3 replied:
fixed now I think?

grodrigues3 commented:
I was going to create a separate plugin for inferring the sig label based on the @ notification, but the code is so similar to the label adding code, I figured I"d just add it here

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+
+	labelMatches := labelRegex.FindAllStringSubmatch(ic.Comment.Body, -1)
+	sigMatches := sigMatcher.FindAllStringSubmatch(ic.Comment.Body, -1)
+	if (labelMatches == nil || len(labelMatches) == 0) && (sigMatches == nil || len(sigMatches) == 0) {
```
Just `len(labelMatches) == 0`, no need to check for `nil`.

grodrigues3 commented:
@k8s-bot bazel test this

apelisse commented on prow/plugins/label/label_test.go:
```diff
@@ -118,10 +118,10 @@ func TestLabel(t *testing.T) {
 		},
 		{
-			name:              "Can't Add Non Existing Label",
+			name:              "Can't Add Non Existent Label Label",
```
nit: label repeated

apelisse commented on prow/plugins/label/label.go:
```diff
@@ -30,6 +30,7 @@ const pluginName = "label"
 
 var (
 	labelRegex       = regexp.MustCompile(`(?m)^/(area|priority|kind)\s*(.*)$`)
+	sigMatcher       = regexp.MustCompile(`(?m)@sig-([\w-]*?)-(?:misc|test-failures|bugs|feature-requests|proposals|pr-reviews|api-reviews)`)
```
`([\w-]*?)`

I think you should force at least one match (empty group is not useful), and it doesn't have to be non-greedy. i.e. `([\w-]+)`

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+	owner := ic.Repo.Owner.Name
+	repo := ic.Repo.Name
+	number := ic.Issue.Number
+
```
You want to skip comments if they are created by `k8s-ci-robot` (use `gc.BotName()`) or if they are being modified or deleted (`ic.Action != "created"`). We can also consider supporting both issue comments and PR opening, as in [the assign plugin](https://github.com/kubernetes/test-infra/blob/master/prow/plugins/assign/assign.go).

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+		return fmt.Errorf("IsMember failed: %v", err)
+	}
+	if !member {
+		return gc.CreateComment(owner, repo, number, plugins.FormatResponse(ic.Comment, "Only kubernetes org members can add labels."))
```
As written, only `ic.Repo.Owner.Name` members can add labels, not k8s members. Prow also hits kubernetes-incubator among a few others.

grodrigues3 replied:
is there an easy way to extend this to other orgs as well?

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+				continue
+			}
+			if !existingLabels.Has(newLabel) {
+				gc.CreateComment(owner, repo, number, fmt.Sprintf(nonExistentLabel, newLabel))
```
Don't ignore the error. Group non-existent labels and make one comment at the end.

spxtr commented on prow/github/client.go:
```diff
@@ -386,6 +386,24 @@ func (c *Client) CreateStatus(org, repo, ref string, s Status) error {
 	return err
 }
 
+func (c *Client) GetLabels(org, repo string) ([]string, error) {
```
This will only return the first 10 labels, right? You'll either need to repeat the call to get them all, or use a different strategy.

grodrigues3 replied:
Why do you say 10? 

[Get Labels](https://developer.github.com/v3/issues/labels/#list-all-labels-for-this-repository) indicates that it returns all for the repo.  Is it due to pagination, in which case the [default](https://developer.github.com/v3/#pagination) is 30 right?

spxtr replied:
Err yes, 30.

grodrigues3 replied:
suggestions for a way of doing this? do i just have to increment the page option and keep getting the labels until nothing is returned? sounds pretty terrible

spxtr replied:
When you call AddLabels it will return a list of labels on the issue. If that list doesn't contain the labels you tried to add then they don't exist on the repo. You can use that instead of a call to GetLabels.

spxtr replied:
This is what we do for the assign plugin. It tries to assign people, but if the return value doesn't have them assigned then it knows they're not an org member.

grodrigues3 replied:
I don't think it works the same way.  AddLabel simply creates a new label.  Assigning people obviously doesn't create new people if they don't exist nor does it assign if member is not part of the org.

spxtr replied:
Ah, AddLabel creates a new label if it doesn't exist? That's annoying.

You can look at `ListIssueComments` to see how to get all labels.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+		return gc.CreateComment(owner, repo, number, plugins.FormatResponse(ic.Comment, "Only kubernetes org members can add labels."))
+	}
+
+	labels, err := gc.GetLabels(owner, repo)
```
Have you checked what GitHub returns if you add labels that don't exist? It might be easier to just attempt to add all labels and then use the return value to figure out which ones weren't added.

grodrigues3 replied:
i have checked this.  Guess which label I was hoping wouldn't get created: [Test PR](https://github.com/grodrigues3/contrib/pull/6).  (TLDR: github just creates labels if they don't exist already)

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+			continue
+		}
+		if !existingLabels.Has(sigLabel) {
+			gc.CreateComment(owner, repo, number, fmt.Sprintf(nonExistentLabel, sigLabel))
```
Don't ignore the error.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	if len(failures) > 0 {
+		gc.CreateComment(owner, repo, number, fmt.Sprintf(labelFailed, strings.Join(failures, ", "), number))
```
Don't ignore the error.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,133 @@
+	if err != nil {
+		return fmt.Errorf("IsMember failed: %v", err)
+	}
+	if !member {
```
Some people have requested allowing anyone to add these labels. Thoughts?

apelisse replied:
Yeah, I'm absolutely not opposed to that ...

grodrigues3 replied:
sure that's fine with me.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,133 @@
+	if err != nil {
+		return err
+	}
+	existingLabels := sets.NewString(labels...)
```
```go
existingLabels := map[string]bool{}
for _, l := range labels {
    existingLabels[l] = true
}
```
And then instead of
```go
if !existingLabels.Has(newLabel) {
```
use
```go
if !existingLabels[newLabel] {
```

spxtr replied:
This gets rid of the dependency on `k8s.io/kubernetes/pkg/util/sets`, which doesn't actually exist anymore. It's been moved to `k8s.io/apimachinery/pkg/util/sets`, but I don't think it's worth using it to save 3 trivial lines.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,133 @@
+		}
+	}
+
+	for _, sig_match := range sigMatches {
```
`sigMatch`, not `sig_match`.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	if len(nonexistent) > 0 {
+		if err := gc.CreateComment(owner, repo, number, fmt.Sprintf(nonExistentLabel, strings.Join(nonexistent, ", "))); err != nil {
```
```go
msg := fmt.Sprintf(nonExistentLabel, strings.Join(nonexistent, ", "))
if err := gc.CreateComment(owner, repo, number, plugins.FormatResponse(ic.Issue, msg)); err != nil {
	log.WithError(err).Errorf("Could not create comment \"%s\".", msg)
}
```
And ditto below.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,133 @@
+		}
+	}
+
+	if len(failures) > 0 {
```
If we get an error adding a label and then post the comment successfully then no error line will show up in prow's logs. We should only get an error adding labels in exceptional cases, such as when GitHub is down. I think in this case it's reasonable to just log the error and not leave a comment.

spxtr commented on prow/github/client.go:
```diff
@@ -386,6 +386,24 @@ func (c *Client) CreateStatus(org, repo, ref string, s Status) error {
+	var res struct {
+		Object []map[string]string `json:"object"`
+	}
+	_, err := c.request(&request{
```
Copy the format used by `ListIssueComments`, we can improve it in a future PR.

grodrigues3 commented:
How are we doing with this?  Are we ready to merge?

spxtr commented on commands.md:
```diff
@@ -6,6 +6,9 @@ Command | Implemented By | Who can run it | Description
 `/unassign [@userA @userB @etc]` | prow [assign](./prow/plugins/assign) | kubernetes org members | Unassigns specified people (or yourself if no one is specified)
+`/area` | prow [label](./prow/plugins/label) | anyone | adds an area/<> label if it exists
+`/kind` | prow [label](./prow/plugins/label) | anyone | adds a kind/<> label if it exists
+`/priority` | prow [label](./prow/plugins/label) | anyone | adds a priority/<> label if it exists
```
sig?

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	if len(failures) > 0 {
+		log.WithError(err).Errorf("Github failed to add the following labels: %s", strings.Join(failures, ", "))
```
What err? I think you can just delete the `failures` variable and `log.WithError(err).Errorf("Failed to add label: %s", sigLabel)` when it happens.

spxtr commented:
One more nit. Run `./bump.sh hook` after rebasing and then this looks good.

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,120 @@
+	for _, l := range labels {
+		existingLabels[l.Name] = true
+	}
+	var failures, nonexistent []string
```
Remove `failures`

spxtr commented on prow/plugins/label/label.go:
```diff
@@ -0,0 +1,120 @@
+				continue
+			}
+			if err := gc.AddLabel(owner, repo, number, newLabel); err != nil {
+				failures = append(failures, newLabel)
```
log the error

spxtr commented on prow/Makefile:
```diff
@@ -15,7 +15,7 @@
 
 
-HOOK_VERSION       = 0.91
+HOOK_VERSION       = 0.92
```
93 now

spxtr commented:
Neat. Hook is bumped once more, then
/lgtm

spxtr commented:
/kind bug

grodrigues3 commented:
@spxtr whoops, meant to ask about deploying this to the main repo?  any rule of thumb before point at k/k?
